R/add_p.R

```diff
@@ -509,11 +428,105 @@ calculate_and_add_test_results <- function(x, include, group = NULL, test.args,
   }
 
   # extending modify_stat_N to new columns
-  x$table_styling$header <- x$table_styling$header |>
-    tidyr::fill(any_of("modify_stat_N"), .direction = "downup")
+  x <- .fill_table_header_modify_stats(x)
 
   # add raw results to `.$card`
   x$cards[[calling_fun]] <- lst_results
+  # print warnings/errors from calculations
+  dplyr::bind_rows(x$cards[[calling_fun]]) |>
+    dplyr::filter(.data$stat_name %in% c("estimate", "std.error", "parameter",
+                                         "statistic", "conf.low", "conf.high", "p.value")) |>
+    cards::print_ard_conditions()
 
   x
 }
+
+.calculate_one_test <- function(df_test_meta_data, variable, data, x, group,
+                                test.args, adj.vars, conf.level,
+                                estimate_fun, pvalue_fun,
+                                continuous_variable = NULL,
+                                apply_fmt_fn = FALSE) {
+  chr_expected_stats <- c("estimate", "std.error", "parameter", "statistic", "conf.low", "conf.high", "p.value")
+  lst_captured_results <-
+    cards::eval_capture_conditions(
+      do.call(
+        what =
+          df_test_meta_data |>
+          dplyr::filter(.data$variable %in% .env$variable) |>
+          dplyr::pull("fun_to_run") %>%
+          getElement(1),
+        args = list(
+          data = data, # most arg names here are data, but `tbl_survfit(x)` is a list of survfit
+          variable = variable,
+          by = x$inputs$by,
+          group = group,
+          type = tryCatch(x$inputs$type[[variable]], error = \(e) NULL), # in tbl_survfit(), the type argument is for transforming the estimate, not the summary type
+          test.args = test.args[[variable]],
+          adj.vars = adj.vars,
+          conf.level = conf.level,
+          tbl = x,
+          continuous_variable = continuous_variable
+        )
+      )
+    )
+
+  # if we captured a warning and the result is an ARD, add the warning to the ARD
+  if (!is.null(lst_captured_results[["warning"]]) &&
+      inherits(lst_captured_results[["result"]], "card")) {
+    lst_captured_results[["result"]][["warning"]] <-
+      lst_captured_results[["result"]][["warning"]] |>
+      map(~c(.x, lst_captured_results[["warning"]]))
+  }
+
+  # if the result is null, replace is with a data frame of of NA
+  if (is.null(lst_captured_results[["result"]])) {
+    lst_captured_results[["result"]] <-
+      rep_named(chr_expected_stats, list(NA)) |>
+      as.data.frame()
+  }
+
+  # check class of returned object. Object must be ARD of class 'card' or a data frame with one row
```

Comment:
So the custom function needs to use cards' model constructor right?

Reply:
the result can either me an ARD structure or a data frame with one row (which is returned by `broom::tidy()`)